File: unions/lib/Lowering.cpp
//===- Lowering.cpp - unions CIR -> LLVM patterns -------------*- C++ -*-===//
//
// Three lowering patterns: union_init, union_tag, union_payload.
// Tagged union layout: struct<(i8 tag, [max_bytes x i8] payload)>.
// Reference: Rust enum lowering, MLIR ConversionPatterns.
//
//===----------------------------------------------------------------------===//
#include "unions/Lowering.h"
#include "unions/Ops.h"
#include "unions/Types.h"

#include "mlir/Conversion/LLVMCommon/Pattern.h"
#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
#include "mlir/Transforms/DialectConversion.h"

using namespace mlir;

namespace {

//===----------------------------------------------------------------------===//
// UnionInitOp -> undef + insertvalue tag + store/load payload bytes
//===----------------------------------------------------------------------===//

struct UnionInitOpLowering : public OpConversionPattern<cir::UnionInitOp> {
  using OpConversionPattern::OpConversionPattern;

  LogicalResult matchAndRewrite(
      cir::UnionInitOp op, OpAdaptor adaptor,
      ConversionPatternRewriter &rewriter) const override {
    auto loc = op.getLoc();
    auto unionType = mlir::cast<cir::TaggedUnionType>(op.getType());
    auto resultType = getTypeConverter()->convertType(op.getType());

    // Find variant index
    auto idx = unionType.getVariantIndex(op.getVariant());

    // Create undef struct
    Value result = rewriter.create<LLVM::UndefOp>(loc, resultType);

    // Insert tag (i8)
    auto tagVal = rewriter.create<LLVM::ConstantOp>(
        loc, rewriter.getI8Type(), rewriter.getI8IntegerAttr(idx));
    result = rewriter.create<LLVM::InsertValueOp>(loc, result, tagVal, 0);

    // Insert payload if present — alloca-store-load bitcast pattern
    if (adaptor.getPayload()) {
      auto payload = adaptor.getPayload();
      auto payloadType = payload.getType();

      // Alloca for the payload, store into it
      auto one = rewriter.create<LLVM::ConstantOp>(
          loc, rewriter.getI64Type(), rewriter.getI64IntegerAttr(1));
      auto payloadPtr = rewriter.create<LLVM::AllocaOp>(
          loc, LLVM::LLVMPointerType::get(rewriter.getContext()),
          payloadType, one);
      rewriter.create<LLVM::StoreOp>(loc, payload, payloadPtr);

      // Load as byte array (reinterpret cast via pointer)
      auto structType = mlir::cast<LLVM::LLVMStructType>(resultType);
      auto byteArrayType = structType.getBody()[1];
      auto loaded = rewriter.create<LLVM::LoadOp>(
          loc, byteArrayType, payloadPtr);
      result = rewriter.create<LLVM::InsertValueOp>(loc, result, loaded, 1);
    }

    rewriter.replaceOp(op, result);
    return success();
  }
};

//===----------------------------------------------------------------------===//
// UnionTagOp -> extractvalue [0]
//===----------------------------------------------------------------------===//

struct UnionTagOpLowering : public OpConversionPattern<cir::UnionTagOp> {
  using OpConversionPattern::OpConversionPattern;

  LogicalResult matchAndRewrite(
      cir::UnionTagOp op, OpAdaptor adaptor,
      ConversionPatternRewriter &rewriter) const override {
    rewriter.replaceOpWithNewOp<LLVM::ExtractValueOp>(
        op, adaptor.getInput(), 0);
    return success();
  }
};

//===----------------------------------------------------------------------===//
// UnionPayloadOp -> extractvalue [1] (byte array) + load as target type
//===----------------------------------------------------------------------===//

struct UnionPayloadOpLowering
    : public OpConversionPattern<cir::UnionPayloadOp> {
  using OpConversionPattern::OpConversionPattern;

  LogicalResult matchAndRewrite(
      cir::UnionPayloadOp op, OpAdaptor adaptor,
      ConversionPatternRewriter &rewriter) const override {
    auto loc = op.getLoc();
    auto resultType = getTypeConverter()->convertType(op.getType());

    // Extract byte array
    auto byteArray = rewriter.create<LLVM::ExtractValueOp>(
        loc, adaptor.getInput(), 1);

    // Alloca-store-load bitcast: store byte array, load as target type
    auto byteArrayType = byteArray.getType();
    auto one = rewriter.create<LLVM::ConstantOp>(
        loc, rewriter.getI64Type(), rewriter.getI64IntegerAttr(1));
    auto ptr = rewriter.create<LLVM::AllocaOp>(
        loc, LLVM::LLVMPointerType::get(rewriter.getContext()),
        byteArrayType, one);
    rewriter.create<LLVM::StoreOp>(loc, byteArray, ptr);
    auto loaded = rewriter.create<LLVM::LoadOp>(loc, resultType, ptr);

    rewriter.replaceOp(op, loaded);
    return success();
  }
};

} // anonymous namespace

void cot::populateUnionsPatterns(RewritePatternSet &patterns,
                                  TypeConverter &typeConverter) {
  patterns.add<
      UnionInitOpLowering,
      UnionTagOpLowering,
      UnionPayloadOpLowering>(typeConverter, patterns.getContext());
}

File: unions/rust/src/lowering.rs
//! Lowers CIR tagged union ops to Cranelift IR instructions.
//!
//! Mirrors C++ unions/Lowering.cpp — 3 patterns:
//! UnionInitOpLowering (tag + store payload bytes),
//! UnionTagOpLowering (load i8 at offset 0),
//! UnionPayloadOpLowering (load from offset 1 as target type).
//!
//! Layout: `{ i8 tag (offset 0), [max_payload_bytes x i8] payload (offset 1) }`
//! Tagged unions are stack-allocated and passed as I64 pointers.

#![cfg(feature = "codegen")]

use cranelift_codegen::ir::{self as clir, types, InstBuilder};

use mlif::codegen::lowering_ctx::LoweringCtx;
use mlif::entity::OpId;
use mlif::ir::context::Context;
use mlif::ir::types::TypeKind;
use mlif::ConstructLowering;

/// Cranelift lowering for the CIR unions construct (3 ops).
pub struct UnionsLowering;

impl ConstructLowering for UnionsLowering {
    fn name(&self) -> &str {
        "unions"
    }

    fn lower_op(&self, op: OpId, lctx: &mut LoweringCtx) -> Result<bool, String> {
        let handled = match lctx.ir[op].name() {
            "cir.union_init" => { lower_union_init(op, lctx)?; true }
            "cir.union_tag" => { lower_union_tag(op, lctx)?; true }
            "cir.union_payload" => { lower_union_payload(op, lctx)?; true }
            _ => false,
        };
        Ok(handled)
    }

    fn map_type(&self, ctx: &Context, ty: mlif::TypeId) -> Option<clir::Type> {
        match ctx.type_kind(ty) {
            TypeKind::Extension(ext) if ext.dialect == "cir" && ext.name == "tagged_union" => {
                Some(types::I64)
            }
            _ => None,
        }
    }
}

/// Compute max payload byte size across all variants.
fn max_payload_size(lctx: &LoweringCtx, union_ty: mlif::TypeId) -> u32 {
    let count = lctx.ext_type_param_count(union_ty);
    let mut max = 0u32;
    for i in 0..count {
        if let Some(vty) = lctx.ext_type_param(union_ty, i) {
            let s = lctx.type_size(vty);
            if s > max { max = s; }
        }
    }
    max.max(1)
}

/// Find variant index by name in the union type.
fn variant_index(lctx: &LoweringCtx, union_ty: mlif::TypeId, variant: &str) -> Result<i64, String> {
    let count = lctx.ext_string_param_count(union_ty);
    for i in 1..count {
        if let Some(name) = lctx.ext_string_param(union_ty, i) {
            if name == variant {
                return Ok((i - 1) as i64);
            }
        }
    }
    Err(format!("unknown variant '{}'", variant))
}

/// Lower `cir.union_init "Variant" %payload?` — alloc + store tag + store payload.
/// C++ equivalent: undef + insertvalue(tag) + alloca-store-load bitcast.
fn lower_union_init(op: OpId, lctx: &mut LoweringCtx) -> Result<(), String> {
    let union_ty = lctx.result_type(op);
    let max_payload = max_payload_size(lctx, union_ty);
    let total = 1 + max_payload;

    let variant = lctx.string_attr(op, "variant")?;
    let tag_idx = variant_index(lctx, union_ty, &variant)?;

    let ptr = lctx.stack_alloc(total);

    // Store tag at offset 0.
    let tag_val = lctx.ins().iconst(types::I8, tag_idx);
    lctx.store_at_offset(tag_val, ptr, 0);

    // Store payload at offset 1 if present.
    if lctx.num_operands(op) > 0 {
        let payload = lctx.operand(op, 0)?;
        lctx.store_at_offset(payload, ptr, 1);
    }

    lctx.set_result(op, ptr);
    Ok(())
}

/// Lower `cir.union_tag %input : i8` — load discriminator tag.
/// C++ equivalent: llvm.extractvalue [0].
fn lower_union_tag(op: OpId, lctx: &mut LoweringCtx) -> Result<(), String> {
    let input = lctx.unary_operand(op)?;
    let r = lctx.load_at_offset(types::I8, input, 0);
    lctx.set_result(op, r);
    Ok(())
}

/// Lower `cir.union_payload "Variant" %input : result_type`.
/// C++ equivalent: extractvalue [1] + alloca-store-load reinterpret cast.
fn lower_union_payload(op: OpId, lctx: &mut LoweringCtx) -> Result<(), String> {
    let input = lctx.unary_operand(op)?;
    let cl_type = lctx.result_cranelift_type(op)?;
    let r = lctx.load_at_offset(cl_type, input, 1);
    lctx.set_result(op, r);
    Ok(())
}
